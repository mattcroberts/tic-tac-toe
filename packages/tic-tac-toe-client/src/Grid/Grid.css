.Grid {
  display: flex;
  flex-direction: column;
  height: 100vh;
  text-align: center;
}

.Grid-row {
  display: flex;
  height: 100%;
  flex-direction: row;
}
