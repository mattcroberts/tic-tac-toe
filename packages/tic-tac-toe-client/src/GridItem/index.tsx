import * as React from "react";

import "./GridItem.css";

export interface IItemState {
    player: string;
}
interface IGridItemProps {
    colN: number;
    rowN: number;
    itemState: IItemState;
    onItemClick: (player: string, x: number, y: number) => void;
}

class GridItem extends React.PureComponent<IGridItemProps, any> {
    constructor(props: IGridItemProps) {
        super(props);
        this.onClick = this.onClick.bind(this);
    }
    public render() {
        return (
            <div className="GridItem" onClick={this.onClick}>
                {this.props.itemState.player}
            </div>
        );
    }
    private onClick() {
        this.props.onItemClick(
            this.props.itemState.player,
            this.props.colN,
            this.props.rowN
        );
    }
}

export default GridItem;
