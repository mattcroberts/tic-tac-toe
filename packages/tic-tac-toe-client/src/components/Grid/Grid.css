.root {
    display: flex;
    flex-direction: column;
    flex: 1;
    text-align: center;
    position: relative;
    user-select: none;
}

.row {
    display: flex;
    flex: 1;
}
