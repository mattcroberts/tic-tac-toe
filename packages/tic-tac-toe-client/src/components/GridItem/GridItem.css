:root {
    --border-color: black;
    --cross-color: black;
    --naught-color: black;
    --finished-color: #69656a;
}

.root {
    display: flex;
    flex-flow: column wrap;
    justify-content: center;
    flex: 1;
    border: 1px solid var(--border-color);
    font-size: 150px;
    height: 29vh;
}

.finished {
    border-color: var(--finished-color);
}

.cross {
    color: var(--cross-color);
}

.cross.finished {
    color: var(--finished-color);
}

.naught {
    color: var(--naught-color);
}

.naught.finished {
    color: var(--finished-color);
}

.top {
    border-top: none;
}
.bottom {
    border-bottom: none;
}
.left {
    border-left: none;
}
.right {
    border-right: none;
}
