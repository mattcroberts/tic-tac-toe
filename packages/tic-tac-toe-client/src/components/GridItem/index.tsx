import classnames from "classnames";
import * as React from "react";

import {
    GridItem as IGridItem,
    Symbol as ISymbol
} from "../../../typings/types";

import style from "./GridItem.css";
import getPlayerSymbol from "../../utils/playerMap";

interface IGridItemProps {
    colN: number;
    rowN: number;
    position: {
        top: boolean;
        bottom: boolean;
        left: boolean;
        right: boolean;
    };
    itemState: IGridItem;
    isFinished: boolean;
    onItemClick: (x: number, y: number) => void;
}

class GridItem extends React.PureComponent<IGridItemProps, any> {
    constructor(props: IGridItemProps) {
        super(props);
        this.onClick = this.onClick.bind(this);
    }

    public render() {
        const {
            position,
            itemState: { player }
        } = this.props;

        return (
            <div
                className={classnames(
                    {
                        [style.finished]: this.props.isFinished
                    },
                    {
                        [style.naught]:
                            player && player.symbol === ISymbol.NAUGHT,
                        [style.cross]: player && player.symbol === ISymbol.CROSS
                    },
                    {
                        [style.root]: true,
                        [style.top]: position.top,
                        [style.bottom]: position.bottom,
                        [style.left]: position.left,
                        [style.right]: position.right
                    }
                )}
                onClick={this.onClick}
            >
                <span>
                    {this.props.itemState.player ? (
                        getPlayerSymbol(this.props.itemState.player.symbol)
                    ) : (
                        <>&nbsp;</>
                    )}
                </span>
            </div>
        );
    }
    private onClick() {
        this.props.onItemClick(this.props.colN, this.props.rowN);
    }
}

export default GridItem;
