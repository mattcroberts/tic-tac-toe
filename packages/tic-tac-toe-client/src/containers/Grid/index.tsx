import * as React from "react";
import { graphql, Subscription } from "react-apollo";
import { withRouter, RouteComponentProps } from "react-router-dom";
import { compose } from "recompose";
import { Grid as IGrid } from "../../../typings/types";
import GridPage from "../../pages/Grid";

import * as EXECUTE_TURN from "./executeTurn.graphql";
import * as GET_GRID from "./getGrid.graphql";
import * as GET_GRID_SUBS from "./getGrid.subscription.graphql";
import ErrorPage from "../../pages/Error";

export interface IProps extends RouteComponentProps<{ playerId: string }> {
    executeTurn: (
        options: {
            variables: {
                id: string;
                playerId: string;
                x: number;
                y: number;
            };
        }
    ) => IGrid;
    isMultiplayer: boolean;
    invitedPlayerId: string;
    data: {
        error: any;
        grid: IGrid;
        loading: boolean;
    };
}

export class GridContainer extends React.Component<IProps> {
    public state = {};

    constructor(props: IProps) {
        super(props);
        this.onItemClick = this.onItemClick.bind(this);
    }

    public render() {
        const {
            invitedPlayerId,
            data: { error, grid, loading }
        } = this.props;

        if (loading) {
            return <p>Loading...</p>;
        }

        if (!grid) {
            return <ErrorPage message="Game not found" />;
        }

        if (!grid.players || grid.players.length < 2) {
            return <p>Player error</p>;
        }

        const controllingPlayer = this.props.isMultiplayer
            ? this.getControllingPlayer(grid, invitedPlayerId)
            : grid.currentPlayer;

        if (!controllingPlayer) {
            return <p>Player not found</p>;
        }

        if (error) {
            return <p>{error.message}</p>;
        }

        return (
            <Subscription
                subscription={GET_GRID_SUBS}
                variables={{ id: grid.id }}
            >
                {({
                    data: { gridUpdated } = { gridUpdated: undefined },
                    loading: subsLoading
                }: {
                    data: { gridUpdated: IGrid | undefined };
                    loading: boolean;
                }) => {
                    const g =
                        subsLoading || gridUpdated === undefined
                            ? grid
                            : gridUpdated;
                    return (
                        <GridPage
                            grid={g.gridItems}
                            currentPlayer={g.currentPlayer}
                            winner={g.winner}
                            isDraw={g.winner === null && g.isFinished}
                            size={g.size}
                            onItemClick={this.onItemClick}
                            gameUrls={g.gameUrls}
                            isMultiplayer={this.props.isMultiplayer}
                            controllingPlayer={controllingPlayer}
                        />
                    );
                }}
            </Subscription>
        );
    }

    private getControllingPlayer(grid: IGrid, invitedPlayerId: string) {
        if (!grid.players || grid.players.length < 2) {
            return null;
        }

        return (
            grid.players.find(p => p !== null && p.id === invitedPlayerId) ||
            grid.players[0]
        );
    }

    private onItemClick(x: number, y: number): void {
        const {
            executeTurn,
            data: { grid }
        } = this.props;

        const controllingPlayer = this.props.isMultiplayer
            ? this.getControllingPlayer(grid, this.props.invitedPlayerId)
            : this.props.data.grid.currentPlayer;

        if (
            grid.gridItems[x][y].player === null &&
            controllingPlayer !== null &&
            controllingPlayer.symbol === grid.currentPlayer.symbol
        ) {
            executeTurn({
                variables: {
                    id: grid.id,
                    playerId: controllingPlayer.id,
                    x,
                    y
                }
            });
        }
    }
}

const enhance = compose<{}, IProps>(
    withRouter,
    graphql(GET_GRID, {
        options: (props: any) => ({
            variables: {
                id: props.match.params.gameId
            }
        })
    }),
    graphql(EXECUTE_TURN, {
        name: "executeTurn"
    })
);

export default enhance(GridContainer);
