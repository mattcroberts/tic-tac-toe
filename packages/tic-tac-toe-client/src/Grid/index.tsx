// import { merge } from "lodash";
import * as React from "react";

import GridItem, { IItemState } from "../GridItem";

import "./Grid.css";

export enum Player {
  CROSS = "X",
  NAUGHT = "0"
}
export interface IGridProps {
  grid: IItemState[][];
  currentPlayer?: Player;
  hasWon?: boolean;
}

interface IGridState {
  grid: IItemState[][];
  currentPlayer: Player;
  hasWon: boolean;
  winner: Player | null;
}

class Grid extends React.Component<IGridProps, IGridState> {
  public static defaultProps = {
    currentPlayer: Player.CROSS,
    hasWon: false
  };
  constructor(props: IGridProps) {
    super(props);
    this.state = {
      currentPlayer: props.currentPlayer as Player,
      grid: props.grid,
      hasWon: props.hasWon as boolean,
      winner: null
    };

    this.hasWon = this.hasWon.bind(this);
    this.onItemClick = this.onItemClick.bind(this);
  }

  public render() {
    return (
      <div className="Grid">
        {this.state.hasWon ? `${this.state.winner} HAS WON` : this.renderGrid()}
      </div>
    );
  }

  private hasWon(player: Player, grid: IItemState[][]): boolean {
    const rowWin = grid.reduce((hasWon, row, i, all) => {
      return (
        hasWon ||
        row.every(({ player: p }, j) => {
          return p === player;
        })
      );
    }, false);

    let colWin = false;
    const forwardDiag: IItemState[] = [];
    const backwardDiag: IItemState[] = [];
    for (let i = 0; i < grid.length; i++) {
      const col: IItemState[] = [];
      forwardDiag.push(grid[i][i]);
      backwardDiag.push(grid[i][grid.length - i - 1]);

      for (let j = 0; j < grid[i].length; j++) {
        col.push(grid[j][i]);
      }

      colWin =
        col[0].player === player &&
        col.every(({ player: p }) => {
          return p === col[0].player;
        });

      if (colWin) {
        break;
      }
    }

    const forwardDiagWin =
      forwardDiag[0].player === player &&
      forwardDiag.every(({ player: p }) => {
        return p === forwardDiag[0].player;
      });

    const backwardDiagWin =
      backwardDiag[0].player === player &&
      backwardDiag.every(({ player: p }) => {
        return p === backwardDiag[0].player;
      });

    return rowWin || colWin || forwardDiagWin || backwardDiagWin;
  }

  private onItemClick(colN: number, rowN: number) {
    if (!this.state.hasWon) {
      this.setState(prevState => {
        const nextGrid = Array.from(prevState.grid);
        nextGrid[colN][rowN] = { player: prevState.currentPlayer };
        const crossHasWon = this.hasWon(Player.CROSS, nextGrid);
        const naughtHasWon = this.hasWon(Player.NAUGHT, nextGrid);

        const newState = {
          ...prevState,
          grid: nextGrid,
          hasWon: crossHasWon || naughtHasWon,
          winner: crossHasWon
            ? Player.CROSS
            : naughtHasWon
              ? Player.NAUGHT
              : null
        };

        if (!crossHasWon && !naughtHasWon) {
          newState.currentPlayer =
            prevState.currentPlayer === Player.CROSS
              ? Player.NAUGHT
              : Player.CROSS;
        }

        return newState;
      });
    }
  }

  private renderGrid() {
    return this.state.grid.map((row, rowN) => (
      <div className="Grid-row" key={`${rowN}`}>
        {row.map((item, colN) => (
          <React.Fragment>
            <span>
              {rowN},
              {colN}
            </span>
            <GridItem
              colN={rowN}
              rowN={colN}
              key={`${colN}-${rowN}`}
              onItemClick={this.onItemClick}
              itemState={item}
            />
          </React.Fragment>
        ))}
      </div>
    ));
  }
}

export default Grid;
