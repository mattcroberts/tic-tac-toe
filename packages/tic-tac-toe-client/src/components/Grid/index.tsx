import * as React from "react";
import {
    GridItem as IGridItem,
    Player as IPlayer
} from "../../../typings/types";
import GridItem from "../GridItem";
import Winner from "../Winner";

import style from "./Grid.css";

export interface IProps {
    id: string;
    grid: IGridItem[][];
    currentPlayer: IPlayer;
    controllingPlayer: IPlayer;
    winner?: IPlayer | null;
    isDraw: boolean;
    size: number;
    executeTurn: ({ variables: ExecuteTurnMutationArgs }) => any;
}

class Grid extends React.Component<IProps> {
    public static defaultProps = {
        isDraw: false,
        winner: null
    };

    constructor(props: IProps) {
        super(props);
        this.onItemClick = this.onItemClick.bind(this);
    }

    public render() {
        return (
            <div className={style.root}>
                {this.renderGrid()}
                {this.props.winner || this.props.isDraw ? (
                    <Winner winner={this.props.winner} />
                ) : null}
            </div>
        );
    }

    public onItemClick(x: number, y: number) {
        const { controllingPlayer, currentPlayer, executeTurn } = this.props;

        if (
            this.props.grid[x][y].player === null &&
            controllingPlayer !== null &&
            controllingPlayer.symbol === this.props.currentPlayer.symbol
        ) {
            executeTurn({
                variables: {
                    id: this.props.id,
                    playerId: currentPlayer.id,
                    x,
                    y
                }
            });
        }
    }

    private renderGrid() {
        return this.props.grid.map((row, rowN) => (
            <div className={style.row} key={`${rowN.toString()}`}>
                {row.map((item, colN) => (
                    <GridItem
                        key={`${rowN}-${colN}`}
                        colN={rowN}
                        rowN={colN}
                        onItemClick={this.onItemClick}
                        itemState={item}
                        isFinished={
                            this.props.winner !== null || this.props.isDraw
                        }
                        position={{
                            bottom: rowN === 2,
                            left: colN === 0,
                            right: colN === 2,
                            top: rowN === 0
                        }}
                    />
                ))}
            </div>
        ));
    }
}

export default Grid;
