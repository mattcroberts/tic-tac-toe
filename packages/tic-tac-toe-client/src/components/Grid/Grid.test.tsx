import * as React from "react";
import { shallow } from "enzyme";
import Grid from ".";
import { createGrid } from "../../../test/utils";
import GridItem from "../GridItem";
import { Symbol as ISymbol } from "../../../typings/types";

describe("Grid", () => {
    const GRID_SIZE = 3;
    const defaultProps = {
        id: "123",
        executeTurn: jest.fn(),
        currentPlayer: {
            id: "123",
            symbol: ISymbol.NAUGHT
        },
        controllingPlayer: {
            id: "456",
            symbol: ISymbol.CROSS
        },
        grid: createGrid(GRID_SIZE),
        size: GRID_SIZE,
        isDraw: false,
        onItemClick: jest.fn()
    };

    beforeEach(() => {
        jest.clearAllMocks();
    });

    it("should render", () => {
        const component = shallow(<Grid {...defaultProps} />);
        expect(component).toMatchSnapshot();
    });

    it("should render grid items", () => {
        const component = shallow(<Grid {...defaultProps} />);

        expect(component.find(GridItem)).toHaveLength(GRID_SIZE * GRID_SIZE);
    });

    it("should render NAUGHT Winner", () => {
        const component = shallow(
            <Grid
                {...defaultProps}
                winner={{ id: "123", symbol: ISymbol.NAUGHT }}
            />
        );

        expect(component).toMatchSnapshot();
    });

    it("should render CROSS Winner", () => {
        const component = shallow(
            <Grid
                {...defaultProps}
                winner={{ id: "123", symbol: ISymbol.NAUGHT }}
            />
        );

        expect(component).toMatchSnapshot();
    });

    it("should render DRAW", () => {
        const component = shallow(<Grid {...defaultProps} isDraw={true} />);

        expect(component).toMatchSnapshot();
    });

    describe("onItemClick", () => {
        it("should executeTurn if square is empty", () => {
            const grid = shallow<Grid>(
                <Grid
                    {...{
                        ...defaultProps,
                        currentPlayer: { id: "789", symbol: ISymbol.CROSS }
                    }}
                />
            );

            grid.instance().onItemClick(1, 2);

            expect(defaultProps.executeTurn).toHaveBeenCalled();
            expect(defaultProps.executeTurn).toHaveBeenCalledWith({
                variables: {
                    id: "123",
                    playerId: "789",
                    x: 1,
                    y: 2
                }
            });
        });

        it("should not executeTurn if square is not empty", () => {
            const props = { ...defaultProps };
            props.grid[1][2] = {
                id: "123",
                player: {
                    id: "123",
                    symbol: ISymbol.NAUGHT
                }
            };
            const grid = shallow<Grid>(<Grid {...props} />);

            grid.instance().onItemClick(1, 2);

            expect(defaultProps.executeTurn).not.toHaveBeenCalled();
        });
    });
});
