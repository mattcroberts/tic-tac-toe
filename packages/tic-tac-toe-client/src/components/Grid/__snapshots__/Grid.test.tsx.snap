// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Grid should render 1`] = `
<div
  className="root"
>
  <div
    className="row"
    key="0"
  >
    <GridItem
      colN={0}
      isFinished={false}
      itemState={
        Object {
          "id": "0_0",
          "player": null,
        }
      }
      key="0-0"
      onItemClick={[Function]}
      position={
        Object {
          "bottom": false,
          "left": true,
          "right": false,
          "top": true,
        }
      }
      rowN={0}
    />
    <GridItem
      colN={0}
      isFinished={false}
      itemState={
        Object {
          "id": "0_1",
          "player": null,
        }
      }
      key="0-1"
      onItemClick={[Function]}
      position={
        Object {
          "bottom": false,
          "left": false,
          "right": false,
          "top": true,
        }
      }
      rowN={1}
    />
    <GridItem
      colN={0}
      isFinished={false}
      itemState={
        Object {
          "id": "0_2",
          "player": null,
        }
      }
      key="0-2"
      onItemClick={[Function]}
      position={
        Object {
          "bottom": false,
          "left": false,
          "right": true,
          "top": true,
        }
      }
      rowN={2}
    />
  </div>
  <div
    className="row"
    key="1"
  >
    <GridItem
      colN={1}
      isFinished={false}
      itemState={
        Object {
          "id": "1_0",
          "player": null,
        }
      }
      key="1-0"
      onItemClick={[Function]}
      position={
        Object {
          "bottom": false,
          "left": true,
          "right": false,
          "top": false,
        }
      }
      rowN={0}
    />
    <GridItem
      colN={1}
      isFinished={false}
      itemState={
        Object {
          "id": "1_1",
          "player": null,
        }
      }
      key="1-1"
      onItemClick={[Function]}
      position={
        Object {
          "bottom": false,
          "left": false,
          "right": false,
          "top": false,
        }
      }
      rowN={1}
    />
    <GridItem
      colN={1}
      isFinished={false}
      itemState={
        Object {
          "id": "1_2",
          "player": null,
        }
      }
      key="1-2"
      onItemClick={[Function]}
      position={
        Object {
          "bottom": false,
          "left": false,
          "right": true,
          "top": false,
        }
      }
      rowN={2}
    />
  </div>
  <div
    className="row"
    key="2"
  >
    <GridItem
      colN={2}
      isFinished={false}
      itemState={
        Object {
          "id": "2_0",
          "player": null,
        }
      }
      key="2-0"
      onItemClick={[Function]}
      position={
        Object {
          "bottom": true,
          "left": true,
          "right": false,
          "top": false,
        }
      }
      rowN={0}
    />
    <GridItem
      colN={2}
      isFinished={false}
      itemState={
        Object {
          "id": "2_1",
          "player": null,
        }
      }
      key="2-1"
      onItemClick={[Function]}
      position={
        Object {
          "bottom": true,
          "left": false,
          "right": false,
          "top": false,
        }
      }
      rowN={1}
    />
    <GridItem
      colN={2}
      isFinished={false}
      itemState={
        Object {
          "id": "2_2",
          "player": null,
        }
      }
      key="2-2"
      onItemClick={[Function]}
      position={
        Object {
          "bottom": true,
          "left": false,
          "right": true,
          "top": false,
        }
      }
      rowN={2}
    />
  </div>
</div>
`;

exports[`Grid should render CROSS Winner 1`] = `
<div
  className="root"
>
  <div
    className="row"
    key="0"
  >
    <GridItem
      colN={0}
      isFinished={true}
      itemState={
        Object {
          "id": "0_0",
          "player": null,
        }
      }
      key="0-0"
      onItemClick={[Function]}
      position={
        Object {
          "bottom": false,
          "left": true,
          "right": false,
          "top": true,
        }
      }
      rowN={0}
    />
    <GridItem
      colN={0}
      isFinished={true}
      itemState={
        Object {
          "id": "0_1",
          "player": null,
        }
      }
      key="0-1"
      onItemClick={[Function]}
      position={
        Object {
          "bottom": false,
          "left": false,
          "right": false,
          "top": true,
        }
      }
      rowN={1}
    />
    <GridItem
      colN={0}
      isFinished={true}
      itemState={
        Object {
          "id": "0_2",
          "player": null,
        }
      }
      key="0-2"
      onItemClick={[Function]}
      position={
        Object {
          "bottom": false,
          "left": false,
          "right": true,
          "top": true,
        }
      }
      rowN={2}
    />
  </div>
  <div
    className="row"
    key="1"
  >
    <GridItem
      colN={1}
      isFinished={true}
      itemState={
        Object {
          "id": "1_0",
          "player": null,
        }
      }
      key="1-0"
      onItemClick={[Function]}
      position={
        Object {
          "bottom": false,
          "left": true,
          "right": false,
          "top": false,
        }
      }
      rowN={0}
    />
    <GridItem
      colN={1}
      isFinished={true}
      itemState={
        Object {
          "id": "1_1",
          "player": null,
        }
      }
      key="1-1"
      onItemClick={[Function]}
      position={
        Object {
          "bottom": false,
          "left": false,
          "right": false,
          "top": false,
        }
      }
      rowN={1}
    />
    <GridItem
      colN={1}
      isFinished={true}
      itemState={
        Object {
          "id": "1_2",
          "player": null,
        }
      }
      key="1-2"
      onItemClick={[Function]}
      position={
        Object {
          "bottom": false,
          "left": false,
          "right": true,
          "top": false,
        }
      }
      rowN={2}
    />
  </div>
  <div
    className="row"
    key="2"
  >
    <GridItem
      colN={2}
      isFinished={true}
      itemState={
        Object {
          "id": "2_0",
          "player": null,
        }
      }
      key="2-0"
      onItemClick={[Function]}
      position={
        Object {
          "bottom": true,
          "left": true,
          "right": false,
          "top": false,
        }
      }
      rowN={0}
    />
    <GridItem
      colN={2}
      isFinished={true}
      itemState={
        Object {
          "id": "2_1",
          "player": null,
        }
      }
      key="2-1"
      onItemClick={[Function]}
      position={
        Object {
          "bottom": true,
          "left": false,
          "right": false,
          "top": false,
        }
      }
      rowN={1}
    />
    <GridItem
      colN={2}
      isFinished={true}
      itemState={
        Object {
          "id": "2_2",
          "player": null,
        }
      }
      key="2-2"
      onItemClick={[Function]}
      position={
        Object {
          "bottom": true,
          "left": false,
          "right": true,
          "top": false,
        }
      }
      rowN={2}
    />
  </div>
  <Winner
    winner={
      Object {
        "id": "123",
        "symbol": "NAUGHT",
      }
    }
  />
</div>
`;

exports[`Grid should render DRAW 1`] = `
<div
  className="root"
>
  <div
    className="row"
    key="0"
  >
    <GridItem
      colN={0}
      isFinished={true}
      itemState={
        Object {
          "id": "0_0",
          "player": null,
        }
      }
      key="0-0"
      onItemClick={[Function]}
      position={
        Object {
          "bottom": false,
          "left": true,
          "right": false,
          "top": true,
        }
      }
      rowN={0}
    />
    <GridItem
      colN={0}
      isFinished={true}
      itemState={
        Object {
          "id": "0_1",
          "player": null,
        }
      }
      key="0-1"
      onItemClick={[Function]}
      position={
        Object {
          "bottom": false,
          "left": false,
          "right": false,
          "top": true,
        }
      }
      rowN={1}
    />
    <GridItem
      colN={0}
      isFinished={true}
      itemState={
        Object {
          "id": "0_2",
          "player": null,
        }
      }
      key="0-2"
      onItemClick={[Function]}
      position={
        Object {
          "bottom": false,
          "left": false,
          "right": true,
          "top": true,
        }
      }
      rowN={2}
    />
  </div>
  <div
    className="row"
    key="1"
  >
    <GridItem
      colN={1}
      isFinished={true}
      itemState={
        Object {
          "id": "1_0",
          "player": null,
        }
      }
      key="1-0"
      onItemClick={[Function]}
      position={
        Object {
          "bottom": false,
          "left": true,
          "right": false,
          "top": false,
        }
      }
      rowN={0}
    />
    <GridItem
      colN={1}
      isFinished={true}
      itemState={
        Object {
          "id": "1_1",
          "player": null,
        }
      }
      key="1-1"
      onItemClick={[Function]}
      position={
        Object {
          "bottom": false,
          "left": false,
          "right": false,
          "top": false,
        }
      }
      rowN={1}
    />
    <GridItem
      colN={1}
      isFinished={true}
      itemState={
        Object {
          "id": "1_2",
          "player": null,
        }
      }
      key="1-2"
      onItemClick={[Function]}
      position={
        Object {
          "bottom": false,
          "left": false,
          "right": true,
          "top": false,
        }
      }
      rowN={2}
    />
  </div>
  <div
    className="row"
    key="2"
  >
    <GridItem
      colN={2}
      isFinished={true}
      itemState={
        Object {
          "id": "2_0",
          "player": null,
        }
      }
      key="2-0"
      onItemClick={[Function]}
      position={
        Object {
          "bottom": true,
          "left": true,
          "right": false,
          "top": false,
        }
      }
      rowN={0}
    />
    <GridItem
      colN={2}
      isFinished={true}
      itemState={
        Object {
          "id": "2_1",
          "player": null,
        }
      }
      key="2-1"
      onItemClick={[Function]}
      position={
        Object {
          "bottom": true,
          "left": false,
          "right": false,
          "top": false,
        }
      }
      rowN={1}
    />
    <GridItem
      colN={2}
      isFinished={true}
      itemState={
        Object {
          "id": "2_2",
          "player": null,
        }
      }
      key="2-2"
      onItemClick={[Function]}
      position={
        Object {
          "bottom": true,
          "left": false,
          "right": true,
          "top": false,
        }
      }
      rowN={2}
    />
  </div>
  <Winner
    winner={null}
  />
</div>
`;

exports[`Grid should render NAUGHT Winner 1`] = `
<div
  className="root"
>
  <div
    className="row"
    key="0"
  >
    <GridItem
      colN={0}
      isFinished={true}
      itemState={
        Object {
          "id": "0_0",
          "player": null,
        }
      }
      key="0-0"
      onItemClick={[Function]}
      position={
        Object {
          "bottom": false,
          "left": true,
          "right": false,
          "top": true,
        }
      }
      rowN={0}
    />
    <GridItem
      colN={0}
      isFinished={true}
      itemState={
        Object {
          "id": "0_1",
          "player": null,
        }
      }
      key="0-1"
      onItemClick={[Function]}
      position={
        Object {
          "bottom": false,
          "left": false,
          "right": false,
          "top": true,
        }
      }
      rowN={1}
    />
    <GridItem
      colN={0}
      isFinished={true}
      itemState={
        Object {
          "id": "0_2",
          "player": null,
        }
      }
      key="0-2"
      onItemClick={[Function]}
      position={
        Object {
          "bottom": false,
          "left": false,
          "right": true,
          "top": true,
        }
      }
      rowN={2}
    />
  </div>
  <div
    className="row"
    key="1"
  >
    <GridItem
      colN={1}
      isFinished={true}
      itemState={
        Object {
          "id": "1_0",
          "player": null,
        }
      }
      key="1-0"
      onItemClick={[Function]}
      position={
        Object {
          "bottom": false,
          "left": true,
          "right": false,
          "top": false,
        }
      }
      rowN={0}
    />
    <GridItem
      colN={1}
      isFinished={true}
      itemState={
        Object {
          "id": "1_1",
          "player": null,
        }
      }
      key="1-1"
      onItemClick={[Function]}
      position={
        Object {
          "bottom": false,
          "left": false,
          "right": false,
          "top": false,
        }
      }
      rowN={1}
    />
    <GridItem
      colN={1}
      isFinished={true}
      itemState={
        Object {
          "id": "1_2",
          "player": null,
        }
      }
      key="1-2"
      onItemClick={[Function]}
      position={
        Object {
          "bottom": false,
          "left": false,
          "right": true,
          "top": false,
        }
      }
      rowN={2}
    />
  </div>
  <div
    className="row"
    key="2"
  >
    <GridItem
      colN={2}
      isFinished={true}
      itemState={
        Object {
          "id": "2_0",
          "player": null,
        }
      }
      key="2-0"
      onItemClick={[Function]}
      position={
        Object {
          "bottom": true,
          "left": true,
          "right": false,
          "top": false,
        }
      }
      rowN={0}
    />
    <GridItem
      colN={2}
      isFinished={true}
      itemState={
        Object {
          "id": "2_1",
          "player": null,
        }
      }
      key="2-1"
      onItemClick={[Function]}
      position={
        Object {
          "bottom": true,
          "left": false,
          "right": false,
          "top": false,
        }
      }
      rowN={1}
    />
    <GridItem
      colN={2}
      isFinished={true}
      itemState={
        Object {
          "id": "2_2",
          "player": null,
        }
      }
      key="2-2"
      onItemClick={[Function]}
      position={
        Object {
          "bottom": true,
          "left": false,
          "right": true,
          "top": false,
        }
      }
      rowN={2}
    />
  </div>
  <Winner
    winner={
      Object {
        "id": "123",
        "symbol": "NAUGHT",
      }
    }
  />
</div>
`;
